### PR TITLE
Refactor RateLimitChecker

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

The current rate limit checker uses hard-coded limits for everything, and the intention of this PR is to make it externally configurable. Additionally the code will be cleaned up a bit.

## Related Tickets & Documents

#5520 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [X] no documentation needed

### CODE REVIEW

**Comment by citizen428** [comment]
This is a WIP/discussion PR for now, I'll post comments inline.

File: app/labor/rate_limit_checker.rb [hunk 1]
```diff
@@ -9,22 +9,13 @@ def initialize(user = nil)
     @user = user
   end
 
-  class UploadRateLimitReached < StandardError; end
-  class DailyFollowAccountLimitReached < StandardError; end
+  UploadRateLimitReached = Class.new(StandardError)
```
Comment (citizen428) [inline app/labor/rate_limit_checker.rb]: Semantically these two forms are the same (`class` is essentially syntactic sugar for creating class objects and assigning them to constants), but the previous one always felt a bit noisy to me for 1 line class definitions.

File: app/labor/rate_limit_checker.rb [hunk 2]
```diff
@@ -9,22 +9,13 @@ def initialize(user = nil)
...
+  DailyFollowAccountLimitReached = Class.new(StandardError)
 
   def limit_by_action(action)
-    result = case action
```
Comment (citizen428) [inline app/labor/rate_limit_checker.rb]: I'm not a big fan of `case` statements like this, so went for an approach with `public_send` and dedicated methods instead. This also means that instead of 

```ruby
RateLimitChecker.new.limit_by_action("comment_creation")
```

we can also do 

```ruby
RateLimitChecker.new.check_comment_creation_limit
```
which is a bit easier for discoverability/code navigation.

Maybe we can even leave off the `_limit` postfix, it seems a bit redundant.

File: app/labor/rate_limit_checker.rb [hunk 3]
```diff
@@ -9,22 +9,13 @@ def initialize(user = nil)
...
-               user.articles.published.where("created_at > ?", 30.seconds.ago).size > 9
-             when "image_upload"
-               Rails.cache.read("#{user.id}_image_upload").to_i > 9
-             when "follow_account"
-               user_today_follow_count > self.class.daily_account_follow_limit
-             else
-               false
-             end
+    check_method = "check_#{action}_limit"
+    result = respond_to?(check_method) ? public_send(check_method) : false
```
Comment (citizen428) [inline app/labor/rate_limit_checker.rb]: Not a big fan of this ternary. Originally I had 

```ruby
result = public_send(check_method) if respond_to?(check_method)
```

but the specs expect `false` explicitly, not a falsey value.  I could have changed them to use the `be_falsey` matcher, but since this is still a WIP PR I decided to wait for feedback on this.

File: app/labor/rate_limit_checker.rb [hunk 4]
```diff
@@ -48,13 +39,26 @@ def ping_admins
...
+
+  def check_follow_account_limit
+    user_today_follow_count > self.class.daily_account_follow_limit
+  end
+
   private
 
   def user_today_follow_count
     following_users_count = user.following_users_count
     return following_users_count if following_users_count < self.class.daily_account_follow_limit
```
Comment (citizen428) [inline app/labor/rate_limit_checker.rb]: I'm not sure why `daily_account_follow_limit` was added as a class method. We are only using it in one place except for this class and its spec:

```ruby
      # spec/requests/follows_create_spec.rb:23
      allow(rate_limit_checker).
        to receive(:user_today_follow_count).
        and_return(RateLimitChecker.daily_account_follow_limit + 1)
```

Given that we could just as well just stub the call on `SiteConfig` directly, I'm not sure it makes sense to keep this class method around at all, we could just inline the `SiteConfig.rate_limit_follow_count_daily` calls and remove the indirection.

File: app/labor/rate_limit_checker.rb (hunk 3 above)
Comment (rhymes) [reply app/labor/rate_limit_checker.rb]: Looks fine

File: app/labor/rate_limit_checker.rb [hunk 5]
```diff
@@ -48,13 +39,26 @@ def ping_admins
...
+    user.articles.published.where("created_at > ?", 30.seconds.ago).size > 9
+  end
+
+  def check_image_upload_limit
+    Rails.cache.read("#{user.id}_image_upload").to_i > 9
+  end
+
+  def check_follow_account_limit
+    user_today_follow_count > self.class.daily_account_follow_limit
+  end
```
Comment (rhymes) [inline app/labor/rate_limit_checker.rb]: why keep them all public? I would even move `ping_admins` in the private section

File: app/labor/rate_limit_checker.rb (hunk 4 above)
Comment (rhymes) [reply app/labor/rate_limit_checker.rb]: I agree

File: app/labor/rate_limit_checker.rb [hunk 6]
```diff
@@ -48,13 +39,26 @@ def ping_admins
...
+
   private
 
   def user_today_follow_count
     following_users_count = user.following_users_count
     return following_users_count if following_users_count < self.class.daily_account_follow_limit
-
-    now = Time.zone.now
-    user.follows.where(created_at: (now.beginning_of_day..now)).size
```
Comment (rhymes) [inline app/labor/rate_limit_checker.rb]: ?

File: app/labor/rate_limit_checker.rb [hunk 7]
```diff
@@ -13,18 +9,9 @@ class UploadRateLimitReached < StandardError; end
...
-               user.articles.published.where("created_at > ?", 30.seconds.ago).size > 9
-             when "image_upload"
-               Rails.cache.read("#{user.id}_image_upload").to_i > 9
-             when "follow_account"
-               user_today_follow_count > self.class.daily_account_follow_limit
-             else
-               false
-             end
+    check_method = "check_#{action}_limit"
+    result = respond_to?(check_method, true) ? send(check_method) : false
```
Comment (citizen428) [inline app/labor/rate_limit_checker.rb]: Note the `true` as second argument to `respond_to?` so it also considers private methods.

File: app/models/site_config.rb [hunk 8]
```diff
@@ -20,6 +20,9 @@ class SiteConfig < RailsSettings::Base
 
   # rate limits
   field :rate_limit_follow_count_daily, type: :integer, default: 500
+  field :rate_limit_comment_creation, type: :integer, default: 9
+  field :rate_limit_published_article_creation, type: :integer, default: 9
+  field :rate_limit_image_upload, type: :integer, default: 9
```
Comment (rhymes) [inline app/models/site_config.rb]: Don't forget to submit another PR after this to expose these rate limits to `/internal/config`

File: spec/labor/rate_limit_checker_spec.rb [hunk 9]
```diff
@@ -4,65 +4,59 @@
...
     it "returns false for invalid action" do
-      expect(described_class.new(user).limit_by_action("random-nothing")).to eq(false)
+      expect(rate_limit_checker.limit_by_action("random-nothing")).to eq(false)
     end
 
     it "returns true if too many comments at once" do
-      create_list(:comment, 10, user_id: user.id, commentable_id: article.id)
-      expect(described_class.new(user).limit_by_action("comment_creation")).to eq(true)
+      count = SiteConfig.rate_limit_comment_creation + 1
+      create_list(:comment, count, user_id: user.id, commentable_id: article.id)
```
Comment (rhymes) [inline spec/labor/rate_limit_checker_spec.rb]: I haven't tested but would it be possible to mock or lower the rate limit as a precondition? Creating 10 comments in the DB when we can create just 2 or 3 it's a waste :D

File: app/models/site_config.rb (hunk 8 above)
Comment (citizen428) [reply app/models/site_config.rb]: I was torn on whether or not to do it in the same PR. In the end my reasoning was that I can finish the refactoring first, since the current version is also non-configurable. Then add the configuration options as separate PR, which has the advantage that the PRs are more limited in scope and therefore easier to review. WDYT?

File: spec/labor/rate_limit_checker_spec.rb (hunk 9 above)
Comment (citizen428) [reply spec/labor/rate_limit_checker_spec.rb]: Good call, reduced the number of objects created.

**Comment by citizen428** [comment]
@rhymes Thanks for the review. I also noted that we're still missing one `SiteConfig` for number of emails to the same recipient, I added that now.

File: app/models/site_config.rb (hunk 8 above)
Comment (rhymes) [reply app/models/site_config.rb]: yeah, that works for me!

File: app/labor/rate_limit_checker.rb [hunk 10]
```diff
@@ -40,19 +27,38 @@ def track_image_uploads
 
   def limit_by_email_recipient_address(address)
     # This is related to the recipient, not the "user" initiator, like in action.
-    EmailMessage.where(to: address).
-      where("sent_at > ?", 2.minutes.ago).size > 5
+    EmailMessage.where(to: address).where("sent_at > ?", 2.minutes.ago).count >
```
Comment (rhymes) [inline app/labor/rate_limit_checker.rb]: Let's switch back this to `.size`, it doesn't hurt, and if email messages are not loaded as they would be it's going to call `.count` anyway

File: app/labor/rate_limit_checker.rb (hunk 7 above)
Comment (mstruve) [reply app/labor/rate_limit_checker.rb]: TIL that `respond_to?` takes a second parameter!
